# Table of Contents
- Import libraries

- Import Data

- Mixed Data Types

- Missing Values

- Duplicates

- Task

- Export Data

# 01. Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# Path Setup
path = r'C:\Users\tom\Documents\Career Foundry\2022-05-29 Instacart Basket Analysis'

In [3]:
# List Setup
vars_list = ['order_id', 'user_id', 'order_number', 'orders_day_of_week', 'order_hour_of_day', 'days_since_last_order']

In [4]:
# Import Products Data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Import Order Data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), usecols = vars_list)

# 03. Mixed-Type Data

In [6]:
# Create a Dataframe
df_test = pd.DataFrame()

In [7]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
df_test['mix'] = df_test['mix'].astype('str')

# 04. Missing Values

In [11]:
# Find columns with missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Create nan subset
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
# View nan subset
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
# View dataframe dimensions
df_prods.shape

(49693, 5)

In [15]:
# Create clean product subset
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
# View dataframe dimensions
df_prods_clean.shape

(49677, 5)

# 05. Duplicates

In [17]:
#Look for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
# View duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
#  New products dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [21]:
# View no duplicates products dataframe
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [22]:
# Export cleaned products dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 06. Task

In [23]:
# 2) View products statistics
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

 The maximum value looks exceptionally high - it is unlikely a product is being sold for $99,999.

In [24]:
# 3) Check for mixed-type data in orders dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There are no mixed-type data columns in df_ords.

In [25]:
# 5) Check for missing values in orders dataframe
df_ords.isnull().sum()

order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [26]:
# Create nan subset for missing values
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [27]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

Lots of orders a missing values in the 'days_since_last_order' column. Looking at the order numbers, it seems most customers have only order once, meaning they would not have a last order.

6) The missing values are valid - they're due to the customer being new, and therefore not having a last order. So instead of doing anything to them, we could create an additional 'new customers' column, providing information as to whether the customer is new or not.

In [28]:
# Create a new orders dataframe
df_ords_clean = df_ords

In [29]:
# Create new customers column
df_ords_clean['new_customer'] = df_ords['days_since_last_order'].isnull() == True

In [30]:
# View new dataframe
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_last_order  new_customer  
0                        8                    NaN          True  
1                        7                   15.0         False  
2                       12                   21.0         False  
3                        7                   29.0         False  
4                       15                   28.0         False  
...                    ...                    ...           ...  
3421078                 18                   29.0         False  
3421079                 10                   30.0         False  
3421080                 12                   18.0         False  
3421081                 12                    7.0         False  
3421082                 14                   30.0         False  

[3421083 rows x 7 columns]

In [31]:
# 7) Create new orders dataframe without duplicates
df_ords_clean_dups = df_ords_clean[df_ords_clean.duplicated()]

In [32]:
df_ords_clean_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order, new_customer]
Index: []

There are no duplicates in the orders dataframe.

In [33]:
df_ords_clean_dups.dtypes

order_id                   int64
user_id                    int64
order_number               int64
orders_day_of_week         int64
order_hour_of_day          int64
days_since_last_order    float64
new_customer                bool
dtype: object

# 07. Export Data

In [34]:
# 9) Export new cleaned orders dataframe
df_ords_clean_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'order_cleaned.csv'))